Error when updating a deliveryservice

I am getting an error when trying to update a deliveryservice from the Traffic Ops UI.  
This error occurs when tried against 1.4.x and master.

```
[2016-02-19 11:54:40,101] [DEBUG] Routing to controller "UI::DeliveryService" and action "update".
Odd number of elements in anonymous hash at lib/UI/DeliveryService.pm line 339.
Use of uninitialized value $typename in string eq at lib/UI/DeliveryService.pm line 343.
Use of uninitialized value $typename in pattern match (m//) at lib/UI/DeliveryService.pm line 574.
Use of uninitialized value in numeric eq (==) at lib/UI/DeliveryService.pm line 909.
[2016-02-19 11:54:40,328] [ERROR] DBIx::Class::Storage::DBI::_dbh_execute(): DBI Exception: DBD::mysql::st execute failed: Cannot add or update a child row: a foreign key constraint fails (`to_development`.`deliveryservice`, CONSTRAINT `fk_deliveryservice_type1` FOREIGN KEY (`type`) REFERENCES `type` (`id`) ON DELETE NO ACTION ON UPDATE NO ACTION) [for Statement "UPDATE deliveryservice SET cacheurl = ?, edge_header_rewrite = ?, global_max_mbps = ?, mid_header_rewrite = ?, origin_shield = ?, regex_remap = ?, remap_text = ?, tr_request_headers = ?, type = ?, xml_id = ? WHERE ( ( id = ? AND type = ? ) )" with ParamValues: 0="", 1="", 2=-1, 3="", 4="", 5="", 6="", 7="", 8=undef, 9="foo", 10='442', 11='8'] at lib/UI/DeliveryService.pm line 928
[2016-02-19 11:54:40,328] [DEBUG] Template "exception.development.html.ep" not found.
Warning: Log message argument #1 undefined at local/lib/perl5/Mojolicious.pm line 199.
Use of uninitialized value in join or string at local/lib/perl5/Log/Log4perl/Appender.pm line 177.
Warning: Log message argument #1 undefined at local/lib/perl5/Mojolicious.pm line 199.
Use of uninitialized value in join or string at local/lib/perl5/Log/Log4perl/Appender.pm line 177.
[2016-02-19 11:54:40,330] [ERROR]
```

--- a/traffic_ops/app/lib/UI/DeliveryService.pm
+++ b/traffic_ops/app/lib/UI/DeliveryService.pm
@@ -28,7 +28,8 @@ sub index {
 	my $self = shift;
 
 	my $pparam = $self->db->resultset('ProfileParameter')->search(
-		{   -and => [
+		{
+			-and => [
 				'parameter.name' => 'deliveryservice_graph_url',
 				'profile.name'   => 'GLOBAL'
 			]
@@ -45,23 +46,16 @@ sub edit {
 	my $self = shift;
 	my $id   = $self->param('id');
 
-	my $rs_ds = $self->db->resultset('Deliveryservice')->search(
-		{ 'me.id' => $id },
-		{   prefetch => [ 'cdn', { 'type' => undef }, { 'profile' => undef } ]
-		}
-	);
+	my $rs_ds =
+		$self->db->resultset('Deliveryservice')->search( { 'me.id' => $id }, { prefetch => [ 'cdn', { 'type' => undef }, { 'profile' => undef } ] } );
 	my $data = $rs_ds->single;
 	my $action;
 	my $regexp_set = &get_regexp_set( $self, $id );
 	my $cdn_domain = &get_cdn_domain( $self, $id );
-	my @example_urls
-		= &get_example_urls( $self, $id, $regexp_set, $data, $cdn_domain,
-		$data->protocol );
+	my @example_urls = &get_example_urls( $self, $id, $regexp_set, $data, $cdn_domain, $data->protocol );
 
-	my $server_count = $self->db->resultset('DeliveryserviceServer')
-		->search( { deliveryservice => $id } )->count();
-	my $static_count = $self->db->resultset('Staticdnsentry')
-		->search( { deliveryservice => $id } )->count();
+	my $server_count = $self->db->resultset('DeliveryserviceServer')->search( { deliveryservice => $id } )->count();
+	my $static_count = $self->db->resultset('Staticdnsentry')->search( { deliveryservice => $id } )->count();
 	&stash_role($self);
 	$self->stash(
 		ds           => $data,
@@ -111,11 +105,7 @@ sub get_example_urls {
 					$url = $scheme . '://edge.' . $host . "." . $cdn_domain;
 					push( @example_urls, $url );
 					if ($scheme2) {
-						$url
-							= $scheme2
-							. '://edge.'
-							. $host . "."
-							. $cdn_domain;
+						$url = $scheme2 . '://edge.' . $host . "." . $cdn_domain;
 						push( @example_urls, $url );
 					}
 				}
@@ -138,30 +128,19 @@ sub get_example_urls {
 				$host =~ s/\\//g;
 				$host =~ s/\.\*//g;
 				$host =~ s/\.//g;
-				my $p
-					= defined( $example_urls[ $re->{set_number} ] )
+				my $p =
+					defined( $example_urls[ $re->{set_number} ] )
 					? $example_urls[ $re->{set_number} ]
 					: "";
 				if ( $re->{set_number} == 0 ) {
-					$example_urls[ $re->{set_number} ]
-						= $scheme
-						. '://ccr.'
-						. $host . "."
-						. $cdn_domain
-						. $p;
+					$example_urls[ $re->{set_number} ] = $scheme . '://ccr.' . $host . "." . $cdn_domain . $p;
 					if ($scheme2) {
-						$url
-							= $scheme2
-							. '://ccr.'
-							. $host . "."
-							. $cdn_domain
-							. $p;
+						$url = $scheme2 . '://ccr.' . $host . "." . $cdn_domain . $p;
 						push( @example_urls, $url );
 					}
 				}
 				else {
-					$example_urls[ $re->{set_number} ]
-						= $scheme . '://' . $re->{pattern} . $p;
+					$example_urls[ $re->{set_number} ] = $scheme . '://' . $re->{pattern} . $p;
 					if ($scheme2) {
 						$url = $scheme2 . '://' . $re->{pattern} . $p;
 						push( @example_urls, $url );
@@ -185,13 +164,9 @@ sub get_cdn_domain {
 	my $self       = shift;
 	my $id         = shift;
 	my $cdn_domain = $self->db->resultset('Parameter')->search(
-		{   -and =>
-				[ 'me.name' => 'domain_name', 'deliveryservices.id' => $id ]
-		},
-		{   join => {
-				profile_parameters =>
-					{ profile => { deliveryservices => undef } }
-			},
+		{ -and => [ 'me.name' => 'domain_name', 'deliveryservices.id' => $id ] },
+		{
+			join     => { profile_parameters => { profile => { deliveryservices => undef } } },
 			distinct => 1
 		}
 	)->get_column('value')->single();
@@ -202,9 +177,8 @@ sub get_regexp_set {
 	my $self = shift;
 	my $id   = shift;
 	my $regexp_set;
-	my $i  = 0;
-	my $rs = $self->db->resultset('RegexesForDeliveryService')
-		->search( {}, { bind => [$id] } );
+	my $i = 0;
+	my $rs = $self->db->resultset('RegexesForDeliveryService')->search( {}, { bind => [$id] } );
 	while ( my $row = $rs->next ) {
 
 		# my $p = $row->pattern;
@@ -224,8 +198,8 @@ sub read {
 	my $orderby = "xml_id";
 	$orderby = $self->param('orderby') if ( defined $self->param('orderby') );
 	my $rs_data = $self->db->resultset("Deliveryservice")->search(
-		undef,
-		{   prefetch => [ 'cdn', 'deliveryservice_regexes' ],
+		undef, {
+			prefetch => [ 'cdn', 'deliveryservice_regexes' ],
 			order_by => 'me.' . $orderby
 		}
 	);
@@ -236,16 +210,16 @@ sub read {
 
 		while ( my $re_row = $re_rs->next ) {
 			push(
-				@matchlist,
-				{   type       => $re_row->regex->type->name,
+				@matchlist, {
+					type       => $re_row->regex->type->name,
 					pattern    => $re_row->regex->pattern,
 					set_number => $re_row->set_number,
 				}
 			);
 		}
 		push(
-			@data,
-			{   "xml_id"                 => $row->xml_id,
+			@data, {
+				"xml_id"                 => $row->xml_id,
 				"display_name"           => $row->display_name,
 				"dscp"                   => $row->dscp,
 				"signed"                 => \$row->signed,
@@ -302,25 +276,15 @@ sub delete {
 		$self->flash( alertmsg => "No can do. Get more privs." );
 	}
 	else {
-		my @regexp_id_list
-			= $self->db->resultset('DeliveryserviceRegex')
-			->search( { deliveryservice => $id } )->get_column('regex')
-			->all();
+		my @regexp_id_list = $self->db->resultset('DeliveryserviceRegex')->search( { deliveryservice => $id } )->get_column('regex')->all();
 
-		my $dsname = $self->db->resultset('Deliveryservice')
-			->search( { id => $id } )->get_column('xml_id')->single();
-		my $delete = $self->db->resultset('Deliveryservice')
-			->search( { id => $id } );
+		my $dsname = $self->db->resultset('Deliveryservice')->search( { id => $id } )->get_column('xml_id')->single();
+		my $delete = $self->db->resultset('Deliveryservice')->search( { id => $id } );
 		$delete->delete();
 
-		my $delete_re = $self->db->resultset('Regex')
-			->search( { id => { -in => \@regexp_id_list } } );
+		my $delete_re = $self->db->resultset('Regex')->search( { id => { -in => \@regexp_id_list } } );
 		$delete_re->delete();
-		&log(
-			$self,
-			"Delete deliveryservice with id:" . $id . " and name " . $dsname,
-			"UICHANGE"
-		);
+		&log( $self, "Delete deliveryservice with id:" . $id . " and name " . $dsname, "UICHANGE" );
 	}
 	return $self->redirect_to('/close_fancybox.html');
 }
@@ -329,33 +293,25 @@ sub check_deliveryservice_input {
 	my $self = shift;
 
 	if ( $self->param('ds.xml_id') =~ /\s/ ) {
-		$self->field('ds.xml_id')
-			->is_equal( "",
-			"Delivery service xml_id cannot contain whitespace." );
+		$self->field('ds.xml_id')->is_equal( "", "Delivery service xml_id cannot contain whitespace." );
 	}
 
 	# TODO:  what restrictions on display_name?
 
-	my $typename
-		= $self->db->resultset('Type')
-		->search( { id => $self->param('ds.type') } )->get_column('name')
-		->single();
+	my $typename = $self->db->resultset('Type')->search( { id => $self->param('ds.type') } )->get_column('name')->single();
 	if ( $typename eq 'ANY_MAP' ) {
-		return
-			$self->valid
-			; # Anything goes for the ANY_MAP, but ds.type is only set on create
+		return $self->valid;    # Anything goes for the ANY_MAP, but ds.type is only set on create
 	}
 
 	if (   $self->param('ds.qstring_ignore') == 2
 		&& $self->param('ds.regex_remap') ne "" )
 	{
-		$self->field('ds.regex_remap')
-			->is_equal( "",
-			"Regex Remap can not be used when qstring_ignore is 2" );
+		$self->field('ds.regex_remap')->is_equal( "", "Regex Remap can not be used when qstring_ignore is 2" );
 	}
 	my $profile_id = $self->param('ds.profile');
 	my $cdn_domain = $self->db->resultset('Parameter')->search(
-		{   'Name'                       => 'domain_name',
+		{
+			'Name'                       => 'domain_name',
 			'Config_file'                => 'CRConfig.json',
 			'profile_parameters.profile' => $profile_id,
 		},
@@ -366,31 +322,20 @@ sub check_deliveryservice_input {
 	my %dbl_check = ();
 	foreach my $param ( $self->param ) {
 		if ( $param =~ /^re_type_(.*)/ ) {
-			my $check_string = $self->param($param) . "|"
-				. $self->param( 're_order_' . $1 );
+			my $check_string = $self->param($param) . "|" . $self->param( 're_order_' . $1 );
 			if ( defined( $dbl_check{$check_string} ) ) {
-				$self->field('ds.regex')
-					->is_equal( "",
-					"Duplicate type/order combination is not allowed." );
+				$self->field('ds.regex')->is_equal( "", "Duplicate type/order combination is not allowed." );
 			}
 			else {
 				$dbl_check{$check_string} = 1;
 			}
-			if (!(     $self->param($param) eq 'HOST_REGEXP'
-					|| $self->param($param) eq 'PATH_REGEXP'
-					|| $self->param($param) eq 'HEADER_REGEXP'
-				)
-				)
-			{
-				$self->field('ds.regex')
-					->is_equal( "",
-					$self->param($param) . " is not a valid regexp type" );
+			if ( !( $self->param($param) eq 'HOST_REGEXP' || $self->param($param) eq 'PATH_REGEXP' || $self->param($param) eq 'HEADER_REGEXP' ) ) {
+				$self->field('ds.regex')->is_equal( "", $self->param($param) . " is not a valid regexp type" );
 			}
 		}
 		elsif ( $param =~ /^re_re_/ ) {
 			if ( $self->param($param) eq "" ) {
-				$self->field('ds.regex')
-					->is_equal( "", "Regular expression cannot be empty." );
+				$self->field('ds.regex')->is_equal( "", "Regular expression cannot be empty." );
 			}
 			else {
 				my $err .= $self->check_regexp( $self->param($param) );
@@ -402,49 +347,32 @@ sub check_deliveryservice_input {
 			if ( $param =~ /^re_re_(\d+)/ || $param =~ /^re_re_new_(\d+)/ ) {
 				my $order_no = $1;
 				my $type_id = &type_id( $self, 'HOST_REGEXP' );
-				if ((   defined( $self->param( 're_type_' . $order_no ) )
-						&& $self->param( 're_type_' . $order_no ) eq
-						'HOST_REGEXP'
-					)
+				if (
+					( defined( $self->param( 're_type_' . $order_no ) ) && $self->param( 're_type_' . $order_no ) eq 'HOST_REGEXP' )
 					|| ( defined( $self->param( 're_type_new_' . $order_no ) )
-						&& $self->param( 're_type_new_' . $order_no ) eq
-						'HOST_REGEXP' )
+						&& $self->param( 're_type_new_' . $order_no ) eq 'HOST_REGEXP' )
 					)
 				{
-					my $new_re
-						= $self->param( 're_re_' . $order_no )
+					my $new_re =
+						  $self->param( 're_re_' . $order_no )
 						? $self->param( 're_re_' . $order_no ) . $cdn_domain
-						: $self->param( 're_re_new_' . $order_no )
-						. $cdn_domain;
-					my $new_order
-						= defined( $self->param( 're_order_' . $order_no ) )
+						: $self->param( 're_re_new_' . $order_no ) . $cdn_domain;
+					my $new_order =
+						defined( $self->param( 're_order_' . $order_no ) )
 						? $self->param( 're_order_' . $order_no )
 						: $self->param( 're_order_new_' . $order_no );
-					my $rs
-						= $self->db->resultset('DeliveryserviceRegex')
-						->search(
-						undef,
-						{   prefetch => [
-								{ regex           => undef },
-								{ deliveryservice => undef }
-							]
-						}
-						);
+					my $rs =
+						$self->db->resultset('DeliveryserviceRegex')->search( undef, { prefetch => [ { regex => undef }, { deliveryservice => undef } ] } );
 					while ( my $row = $rs->next ) {
 						my $existing_re = $row->regex->pattern . $cdn_domain;
 						if ( defined( $self->param('id') )
-							&& $self->param('id')
-							== $row->deliveryservice->id )
+							&& $self->param('id') == $row->deliveryservice->id )
 						{
 							next;
 						}
 						if ( $existing_re eq $new_re ) {
-							$self->field('ds.regex')->is_equal( "",
-								      "There already is a HOST_REGEXP ("
-									. $existing_re
-									. ") that maches "
-									. $new_re
-									. "; No can do." );
+							$self->field('ds.regex')
+								->is_equal( "", "There already is a HOST_REGEXP (" . $existing_re . ") that maches " . $new_re . "; No can do." );
 							last;
 						}
 					}
@@ -453,9 +381,7 @@ sub check_deliveryservice_input {
 		}
 		elsif ( $param =~ /^re_order_.*(\d+)/ ) {
 			if ( $self->param($param) !~ /^\d+$/ ) {
-				$self->field('ds.regex')
-					->is_equal( "",
-					$self->param($param) . " is not a valid order number." );
+				$self->field('ds.regex')->is_equal( "", $self->param($param) . " is not a valid order number." );
 			}
 		}
 		if ( $self->param($param) eq 'HOST_REGEXP' ) {
@@ -476,131 +402,99 @@ sub check_deliveryservice_input {
 		}
 	}
 	if ( !$match_one ) {
-		$self->field('ds.regex')->is_equal( "",
-			"A minimum of one host regexp with order 0 is needed per delivery service."
-		);
+		$self->field('ds.regex')->is_equal( "", "A minimum of one host regexp with order 0 is needed per delivery service." );
 	}
 	if ( $self->param('ds.dscp') !~ /^\d+$/ ) {
-		$self->field('ds.dscp')
-			->is_equal( "",
-			$self->param('ds.dscp') . " is not a valid dscp value." );
+		$self->field('ds.dscp')->is_equal( "", $self->param('ds.dscp') . " is not a valid dscp value." );
 	}
 
 	my $org_host_name = $self->param('ds.org_server_fqdn');
-	$self->field('ds.org_server_fqdn')
-		->is_like( qr/^(https?:\/\/)/,
-		"Origin Server Base URL must start with http(s)://" );
+	$self->field('ds.org_server_fqdn')->is_like( qr/^(https?:\/\/)/, "Origin Server Base URL must start with http(s)://" );
 	$org_host_name =~ s!^https?://?!!i;
 	$org_host_name =~ s/:(.*)$//;
 	my $port = defined($1) ? $1 : 80;
 	if ( !&is_hostname($org_host_name) || $port !~ /^[1-9][0-9]*$/ ) {
-		$self->field('ds.org_server_fqdn')->is_equal( "",
-			      $org_host_name
-				. " is not a valid org server name (rfc1123) or "
-				. $port
-				. " is not a valid port" );
+		$self->field('ds.org_server_fqdn')
+			->is_equal( "", $org_host_name . " is not a valid org server name (rfc1123) or " . $port . " is not a valid port" );
 	}
 	if ( $self->param('ds.http_bypass_fqdn') ne ""
 		&& !&is_hostname( $self->param('ds.http_bypass_fqdn') ) )
 	{
-		$self->field('ds.http_bypass_fqdn')->is_equal( "",
-			      "Invalid HTTP bypass FQDN "
-				. $self->param('ds.http_bypass_fqdn')
-				. "  : should by FQDN only, not URL. Example: host.overflowcdn.com"
-		);
+		$self->field('ds.http_bypass_fqdn')
+			->is_equal( "",
+			"Invalid HTTP bypass FQDN " . $self->param('ds.http_bypass_fqdn') . "  : should by FQDN only, not URL. Example: host.overflowcdn.com" );
 	}
 	my $dns_bypass_ttl_required;
 	if ( $self->param('ds.dns_bypass_ip') ne "" ) {
 		if ( !&is_ipaddress( $self->param('ds.dns_bypass_ip') ) ) {
-			$self->field('ds.dns_bypass_ip')->is_equal( "",
-				      "DNS bypass IP "
-					. $self->param('ds.dns_bypass_ip')
-					. " is not valid IPv4 address." );
+			$self->field('ds.dns_bypass_ip')->is_equal( "", "DNS bypass IP " . $self->param('ds.dns_bypass_ip') . " is not valid IPv4 address." );
 		}
 		$dns_bypass_ttl_required = 1;
 	}
 	if ( $self->param('ds.dns_bypass_ip6') ne "" ) {
 		if ( !&is_ip6address( $self->param('ds.dns_bypass_ip6') ) ) {
-			$self->field('ds.dns_bypass_ip6')->is_equal( "",
-				      "DNS bypass IPv6 IP ="
-					. $self->param('ds.dns_bypass_ip6')
-					. " is not a valid IPv6 address." );
+			$self->field('ds.dns_bypass_ip6')->is_equal( "", "DNS bypass IPv6 IP =" . $self->param('ds.dns_bypass_ip6') . " is not a valid IPv6 address." );
 		}
 		$dns_bypass_ttl_required = 1;
 	}
 	if ( $self->param('ds.dns_bypass_cname') ne ""
 		&& !&is_hostname( $self->param('ds.dns_bypass_cname') ) )
 	{
-		$self->field('ds.dns_bypass_cname')->is_equal( "",
-			      "Invalid DNS bypass CNAME "
-				. $self->param('ds.dns_bypass_cname')
-				. "  : should by FQDN only, not URL. Example: host.bypass.com"
-		);
+		$self->field('ds.dns_bypass_cname')
+			->is_equal( "",
+			"Invalid DNS bypass CNAME " . $self->param('ds.dns_bypass_cname') . "  : should by FQDN only, not URL. Example: host.bypass.com" );
 		$dns_bypass_ttl_required = 1;
 	}
 	if ( $dns_bypass_ttl_required
 		&& ( $self->param('ds.dns_bypass_ttl') eq "" ) )
 	{
-		$self->field('ds.dns_bypass_ttl')
-			->is_equal( "",
-			"DNS bypass TTL required when specifying DNS bypass IP" );
+		$self->field('ds.dns_bypass_ttl')->is_equal( "", "DNS bypass TTL required when specifying DNS bypass IP" );
 	}
 	if ( defined( $self->param('ds.dns_bypass_ttl') )
 		&& $self->param('ds.dns_bypass_ttl') =~ m/[a-zA-Z]/ )
 	{
-		$self->field('ds.dns_bypass_ttl')->is_equal( "",
-			      "DNS bypass TTL "
-				. $self->param('ds.dns_bypass_ttl')
-				. " should be integers only." );
+		$self->field('ds.dns_bypass_ttl')->is_equal( "", "DNS bypass TTL " . $self->param('ds.dns_bypass_ttl') . " should be integers only." );
 	}
 	if (   defined( $self->param('ds.global_max_mbps') )
 		&& $self->param('ds.global_max_mbps') ne ""
 		&& $self->param('ds.global_max_mbps') !~ /^\d+$/ )
 	{
-		if ( $self->hr_string_to_mbps( $self->param('ds.global_max_mbps') )
-			< 0 )
-		{
-			$self->field('ds.global_max_mbps')
-				->is_equal( "", "Invalid global_max_mbps (NaN)." );
+		if ( $self->hr_string_to_mbps( $self->param('ds.global_max_mbps') ) < 0 ) {
+			$self->field('ds.global_max_mbps')->is_equal( "", "Invalid global_max_mbps (NaN)." );
 		}
 	}
 	if (   $self->param('ds.global_max_tps') ne ""
 		&& $self->param('ds.global_max_tps') !~ /^\d+$/ )
 	{
-		$self->field('ds.global_max_tps')
-			->is_equal( "", "Invalid global_max_tps (NaN)." );
+		$self->field('ds.global_max_tps')->is_equal( "", "Invalid global_max_tps (NaN)." );
 	}
 
 	if ( $typename =~ /^DNS/ ) {
 		if ( defined( $self->param('ds.tr_response_headers') )
 			&& $self->param('ds.tr_response_headers') ne "" )
 		{
-			$self->field('ds.tr_response_headers')->is_equal( "",
-				"TR Response Headers are only valid for HTTP (302) delivery services"
-			);
+			$self->field('ds.tr_response_headers')->is_equal( "", "TR Response Headers are only valid for HTTP (302) delivery services" );
 		}
 		if ( defined( $self->param('ds.tr_request_headers') )
 			&& $self->param('ds.tr_request_headers') ne "" )
 		{
-			$self->field('ds.tr_request_headers')->is_equal( "",
-				"TR Log Request Headers are only valid for HTTP (302) delivery services"
-			);
+			$self->field('ds.tr_request_headers')->is_equal( "", "TR Log Request Headers are only valid for HTTP (302) delivery services" );
 		}
 	}
 
-#TODO:  Fix this to work the right way.
-# if ( defined( $self->param('ds.edge_header_rewrite') ) ) {
-# 	if ( $self->param('ds.edge_header_rewrite') ne "" && $self->param('ds.edge_header_rewrite') !~ /^(?:add|rm|set)-header .* \[L\]$/ ) {
-# 		$self->field('ds.edge_header_rewrite')
-# 			->is_equal( "",
-# 			"edge_header_rewrite is a single line that needs to start with [add|rm|set]-header, and end with [L] - see header rewrite docs." );
-# 	}
-# }
-# if ( defined( $self->param('ds.ipv6_routing_enabled') ) ) {
-# 	if ( $self->param('ds.type.name') =~ /^DNS/ && $self->param('ds.ipv6_routing_enabled') == 0 ) {
-# 		$self->field('ds.ipv6_routing_enabled')->is_equal( "", "IPv6 Routing cannot be disabled for DNS deliveryservices." );
-# 	}
-# }
+	#TODO:  Fix this to work the right way.
+	# if ( defined( $self->param('ds.edge_header_rewrite') ) ) {
+	# 	if ( $self->param('ds.edge_header_rewrite') ne "" && $self->param('ds.edge_header_rewrite') !~ /^(?:add|rm|set)-header .* \[L\]$/ ) {
+	# 		$self->field('ds.edge_header_rewrite')
+	# 			->is_equal( "",
+	# 			"edge_header_rewrite is a single line that needs to start with [add|rm|set]-header, and end with [L] - see header rewrite docs." );
+	# 	}
+	# }
+	# if ( defined( $self->param('ds.ipv6_routing_enabled') ) ) {
+	# 	if ( $self->param('ds.type.name') =~ /^DNS/ && $self->param('ds.ipv6_routing_enabled') == 0 ) {
+	# 		$self->field('ds.ipv6_routing_enabled')->is_equal( "", "IPv6 Routing cannot be disabled for DNS deliveryservices." );
+	# 	}
+	# }
 	return $self->valid;
 }
 
@@ -627,21 +521,15 @@ sub header_rewrite {
 		if ( $tier eq "mid" ) {
 			$fname = "hdr_rw_mid_" . $ds_name . ".config";
 		}
-		my $ats_cfg_loc
-			= $self->db->resultset('Parameter')
-			->search(
-			{ -and => [ name => 'location', config_file => 'remap.config' ] }
-			)->get_column('value')->single();
+		my $ats_cfg_loc =
+			$self->db->resultset('Parameter')->search( { -and => [ name => 'location', config_file => 'remap.config' ] } )->get_column('value')->single();
 		$ats_cfg_loc =~ s/\/$//;
 
-		my $param_id
-			= $self->db->resultset('Parameter')
-			->search(
-			{ -and => [ name => 'location', config_file => $fname ] } )
-			->get_column('id')->single();
+		my $param_id = $self->db->resultset('Parameter')->search( { -and => [ name => 'location', config_file => $fname ] } )->get_column('id')->single();
 		if ( !defined($param_id) ) {
 			my $insert = $self->db->resultset('Parameter')->create(
-				{   config_file => $fname,
+				{
+					config_file => $fname,
 					name        => 'location',
 					value       => $ats_cfg_loc
 				}
@@ -650,46 +538,30 @@ sub header_rewrite {
 			$param_id = $insert->id;
 		}
 		my $cdn_name = undef;
-		my @servers
-			= $self->db->resultset('DeliveryserviceServer')
-			->search( { deliveryservice => $ds_id } )->get_column('server')
-			->all();
+		my @servers = $self->db->resultset('DeliveryserviceServer')->search( { deliveryservice => $ds_id } )->get_column('server')->all();
 		if ( $tier eq "mid" ) {
 			my $mtype_id = &type_id( $self, 'MID' );
-			my $param
-				= $self->db->resultset('Deliveryservice')
-				->search( { 'me.profile' => $ds_profile },
-				{ prefetch => 'cdn' } );
+			my $param = $self->db->resultset('Deliveryservice')->search( { 'me.profile' => $ds_profile }, { prefetch => 'cdn' } );
 			$cdn_name = $param->next->cdn->name;
 
-			@servers = $self->db->resultset('Server')
-				->search( { type => $mtype_id } )->get_column('id')->all();
+			@servers = $self->db->resultset('Server')->search( { type => $mtype_id } )->get_column('id')->all();
 		}
-		my @profiles
-			= $self->db->resultset('Server')
-			->search( { id => { -in => \@servers } } )->get_column('profile')
-			->all();
+		my @profiles = $self->db->resultset('Server')->search( { id => { -in => \@servers } } )->get_column('profile')->all();
 		foreach my $profile_id (@profiles) {
-			my $link
-				= $self->db->resultset('ProfileParameter')
-				->search( { profile => $profile_id, parameter => $param_id } )
-				->single();
+			my $link = $self->db->resultset('ProfileParameter')->search( { profile => $profile_id, parameter => $param_id } )->single();
 			if ( !defined($link) ) {
 				if ($cdn_name) {
-					my $p_cdn_param
-						= $self->db->resultset('Server')
-						->search( { 'me.profile' => $profile_id },
-						{ prefetch => 'cdn' } );
+					my $p_cdn_param = $self->db->resultset('Server')->search( { 'me.profile' => $profile_id }, { prefetch => 'cdn' } );
 					if ( $p_cdn_param->next->cdn->name ne $cdn_name ) {
 						next;
 					}
 				}
-				my $insert
-					= $self->db->resultset('ProfileParameter')->create(
-					{   profile   => $profile_id,
+				my $insert = $self->db->resultset('ProfileParameter')->create(
+					{
+						profile   => $profile_id,
 						parameter => $param_id
 					}
-					);
+				);
 
 			}
 		}
@@ -700,8 +572,7 @@ sub header_rewrite {
 			$fname = "hdr_rw_mid_" . $ds_name . ".config";
 		}
 
-		&delete_cfg_file( $self, $fname )
-			; # don't change it to $self->delete_header_rewrite(), calling from other pm is wonky
+		&delete_cfg_file( $self, $fname );    # don't change it to $self->delete_header_rewrite(), calling from other pm is wonky
 	}
 }
 
@@ -714,21 +585,15 @@ sub regex_remap {
 
 	if ( defined($regex_remap) && $regex_remap ne "" ) {
 		my $fname = "regex_remap_" . $ds_name . ".config";
-		my $ats_cfg_loc
-			= $self->db->resultset('Parameter')
-			->search(
-			{ -and => [ name => 'location', config_file => 'remap.config' ] }
-			)->get_column('value')->single();
+		my $ats_cfg_loc =
+			$self->db->resultset('Parameter')->search( { -and => [ name => 'location', config_file => 'remap.config' ] } )->get_column('value')->single();
 		$ats_cfg_loc =~ s/\/$//;
 
-		my $param_id
-			= $self->db->resultset('Parameter')
-			->search(
-			{ -and => [ name => 'location', config_file => $fname ] } )
-			->get_column('id')->single();
+		my $param_id = $self->db->resultset('Parameter')->search( { -and => [ name => 'location', config_file => $fname ] } )->get_column('id')->single();
 		if ( !defined($param_id) ) {
 			my $insert = $self->db->resultset('Parameter')->create(
-				{   config_file => $fname,
+				{
+					config_file => $fname,
 					name        => 'location',
 					value       => $ats_cfg_loc
 				}
@@ -736,32 +601,23 @@ sub regex_remap {
 			$insert->insert();
 			$param_id = $insert->id;
 		}
-		my @servers
-			= $self->db->resultset('DeliveryserviceServer')
-			->search( { deliveryservice => $ds_id } )->get_column('server')
-			->all();
-		my @profiles
-			= $self->db->resultset('Server')
-			->search( { id => { -in => \@servers } } )->get_column('profile')
-			->all();
+		my @servers = $self->db->resultset('DeliveryserviceServer')->search( { deliveryservice => $ds_id } )->get_column('server')->all();
+		my @profiles = $self->db->resultset('Server')->search( { id => { -in => \@servers } } )->get_column('profile')->all();
 		foreach my $profile_id (@profiles) {
-			my $link
-				= $self->db->resultset('ProfileParameter')
-				->search( { profile => $profile_id, parameter => $param_id } )
-				->single();
+			my $link = $self->db->resultset('ProfileParameter')->search( { profile => $profile_id, parameter => $param_id } )->single();
 			if ( !defined($link) ) {
-				my $insert
-					= $self->db->resultset('ProfileParameter')->create(
-					{   profile   => $profile_id,
+				my $insert = $self->db->resultset('ProfileParameter')->create(
+					{
+						profile   => $profile_id,
 						parameter => $param_id
 					}
-					);
+				);
 			}
 		}
 	}
 	else {
 		&delete_cfg_file( $self, "regex_remap_" . $ds_name . ".config" )
-			; # don't change it to $self->delete_header_rewrite(), calling from other pm is wonky
+			;    # don't change it to $self->delete_header_rewrite(), calling from other pm is wonky
 	}
 }
 
@@ -775,21 +631,15 @@ sub cacheurl {
 
 	if ( defined($cacheurl) && $cacheurl ne "" ) {
 		my $fname = "cacheurl_" . $ds_name . ".config";
-		my $ats_cfg_loc
-			= $self->db->resultset('Parameter')
-			->search(
-			{ -and => [ name => 'location', config_file => 'remap.config' ] }
-			)->get_column('value')->single();
+		my $ats_cfg_loc =
+			$self->db->resultset('Parameter')->search( { -and => [ name => 'location', config_file => 'remap.config' ] } )->get_column('value')->single();
 		$ats_cfg_loc =~ s/\/$//;
 
-		my $param_id
-			= $self->db->resultset('Parameter')
-			->search(
-			{ -and => [ name => 'location', config_file => $fname ] } )
-			->get_column('id')->single();
+		my $param_id = $self->db->resultset('Parameter')->search( { -and => [ name => 'location', config_file => $fname ] } )->get_column('id')->single();
 		if ( !defined($param_id) ) {
 			my $insert = $self->db->resultset('Parameter')->create(
-				{   config_file => $fname,
+				{
+					config_file => $fname,
 					name        => 'location',
 					value       => $ats_cfg_loc
 				}
@@ -797,32 +647,22 @@ sub cacheurl {
 			$insert->insert();
 			$param_id = $insert->id;
 		}
-		my @servers
-			= $self->db->resultset('DeliveryserviceServer')
-			->search( { deliveryservice => $ds_id } )->get_column('server')
-			->all();
-		my @profiles
-			= $self->db->resultset('Server')
-			->search( { id => { -in => \@servers } } )->get_column('profile')
-			->all();
+		my @servers = $self->db->resultset('DeliveryserviceServer')->search( { deliveryservice => $ds_id } )->get_column('server')->all();
+		my @profiles = $self->db->resultset('Server')->search( { id => { -in => \@servers } } )->get_column('profile')->all();
 		foreach my $profile_id (@profiles) {
-			my $link
-				= $self->db->resultset('ProfileParameter')
-				->search( { profile => $profile_id, parameter => $param_id } )
-				->single();
+			my $link = $self->db->resultset('ProfileParameter')->search( { profile => $profile_id, parameter => $param_id } )->single();
 			if ( !defined($link) ) {
-				my $insert
-					= $self->db->resultset('ProfileParameter')->create(
-					{   profile   => $profile_id,
+				my $insert = $self->db->resultset('ProfileParameter')->create(
+					{
+						profile   => $profile_id,
 						parameter => $param_id
 					}
-					);
+				);
 			}
 		}
 	}
 	else {
-		&delete_cfg_file( $self, "cacheurl_" . $ds_name . ".config" )
-			; # don't change it to $self->delete_header_rewrite(), calling from other pm is wonky
+		&delete_cfg_file( $self, "cacheurl_" . $ds_name . ".config" );   # don't change it to $self->delete_header_rewrite(), calling from other pm is wonky
 	}
 }
 
@@ -830,14 +670,10 @@ sub delete_cfg_file {
 	my $self  = shift;
 	my $fname = shift;
 
-	my $param_id
-		= $self->db->resultset('Parameter')
-		->search( { -and => [ name => 'location', config_file => $fname ] } )
-		->get_column('id')->single();
+	my $param_id = $self->db->resultset('Parameter')->search( { -and => [ name => 'location', config_file => $fname ] } )->get_column('id')->single();
 	if ( defined($param_id) ) {
 		$self->app->log->info( 'deleting location parameter for ' . $fname );
-		my $delete = $self->db->resultset('Parameter')
-			->search( { id => $param_id } );
+		my $delete = $self->db->resultset('Parameter')->search( { id => $param_id } );
 		$delete->delete();
 	}
 }
@@ -855,66 +691,53 @@ sub update {
 	}
 	if ( $self->check_deliveryservice_input() ) {
 
-	  #print "global_max_mbps = " . $self->param('ds.global_max_mbps') . "\n";
-	  # if error check passes
+		#print "global_max_mbps = " . $self->param('ds.global_max_mbps') . "\n";
+		# if error check passes
 		my %hash = (
-			xml_id            => $self->paramAsScalar('ds.xml_id'),
-			display_name      => $self->paramAsScalar('ds.display_name'),
-			dscp              => $self->paramAsScalar('ds.dscp'),
-			signed            => $self->paramAsScalar('ds.signed'),
-			qstring_ignore    => $self->paramAsScalar('ds.qstring_ignore'),
-			geo_limit         => $self->paramAsScalar('ds.geo_limit'),
-			org_server_fqdn   => $self->paramAsScalar('ds.org_server_fqdn'),
-			multi_site_origin => $self->paramAsScalar('ds.multi_site_origin'),
-			ccr_dns_ttl       => $self->paramAsScalar('ds.ccr_dns_ttl'),
-			type              => $self->paramAsScalar('ds.type'),
-			cdn_id            => $self->paramAsScalar('ds.cdn_id'),
-			profile           => $self->paramAsScalar('ds.profile'),
-			global_max_mbps   => $self->hr_string_to_mbps(
-				$self->paramAsScalar( 'ds.global_max_mbps', 0 )
-			),
-			global_max_tps => $self->paramAsScalar( 'ds.global_max_tps', 0 ),
-			miss_lat       => $self->paramAsScalar('ds.miss_lat'),
-			miss_long      => $self->paramAsScalar('ds.miss_long'),
-			long_desc      => $self->paramAsScalar('ds.long_desc'),
-			long_desc_1    => $self->paramAsScalar('ds.long_desc_1'),
-			long_desc_2    => $self->paramAsScalar('ds.long_desc_2'),
-			info_url       => $self->paramAsScalar('ds.info_url'),
-			check_path     => $self->paramAsScalar('ds.check_path'),
-			active         => $self->paramAsScalar('ds.active'),
-			protocol       => $self->paramAsScalar('ds.protocol'),
-			ipv6_routing_enabled =>
-				$self->paramAsScalar('ds.ipv6_routing_enabled'),
-			regional_geo_blocking => 
-				$self->paramAsScalar('ds.regional_geo_blocking'),
-			range_request_handling =>
-				$self->paramAsScalar('ds.range_request_handling'),
-			edge_header_rewrite =>
-				$self->paramAsScalar( 'ds.edge_header_rewrite', undef ),
-			mid_header_rewrite =>
-				$self->paramAsScalar( 'ds.mid_header_rewrite', undef ),
-			tr_response_headers =>
-				$self->paramAsScalar( 'ds.tr_response_headers', undef ),
-			tr_request_headers =>
-				$self->paramAsScalar( 'ds.tr_request_headers', undef ),
-			regex_remap => $self->paramAsScalar( 'ds.regex_remap', undef ),
-			origin_shield =>
-				$self->paramAsScalar( 'ds.origin_shield', undef ),
-			cacheurl   => $self->paramAsScalar( 'ds.cacheurl',   undef ),
-			remap_text => $self->paramAsScalar( 'ds.remap_text', undef ),
-			initial_dispersion =>
-				$self->paramAsScalar( 'ds.initial_dispersion', 1 ),
+			xml_id                 => $self->paramAsScalar('ds.xml_id'),
+			display_name           => $self->paramAsScalar('ds.display_name'),
+			dscp                   => $self->paramAsScalar('ds.dscp'),
+			signed                 => $self->paramAsScalar('ds.signed'),
+			qstring_ignore         => $self->paramAsScalar('ds.qstring_ignore'),
+			geo_limit              => $self->paramAsScalar('ds.geo_limit'),
+			org_server_fqdn        => $self->paramAsScalar('ds.org_server_fqdn'),
+			multi_site_origin      => $self->paramAsScalar('ds.multi_site_origin'),
+			ccr_dns_ttl            => $self->paramAsScalar('ds.ccr_dns_ttl'),
+			type                   => $self->paramAsScalar('ds.type'),
+			cdn_id                 => $self->paramAsScalar('ds.cdn_id'),
+			profile                => $self->paramAsScalar('ds.profile'),
+			global_max_mbps        => $self->hr_string_to_mbps( $self->paramAsScalar( 'ds.global_max_mbps', 0 ) ),
+			global_max_tps         => $self->paramAsScalar( 'ds.global_max_tps', 0 ),
+			miss_lat               => $self->paramAsScalar('ds.miss_lat'),
+			miss_long              => $self->paramAsScalar('ds.miss_long'),
+			long_desc              => $self->paramAsScalar('ds.long_desc'),
+			long_desc_1            => $self->paramAsScalar('ds.long_desc_1'),
+			long_desc_2            => $self->paramAsScalar('ds.long_desc_2'),
+			info_url               => $self->paramAsScalar('ds.info_url'),
+			check_path             => $self->paramAsScalar('ds.check_path'),
+			active                 => $self->paramAsScalar('ds.active'),
+			protocol               => $self->paramAsScalar('ds.protocol'),
+			ipv6_routing_enabled   => $self->paramAsScalar('ds.ipv6_routing_enabled'),
+			regional_geo_blocking  => $self->paramAsScalar('ds.regional_geo_blocking'),
+			range_request_handling => $self->paramAsScalar('ds.range_request_handling'),
+			edge_header_rewrite    => $self->paramAsScalar( 'ds.edge_header_rewrite', undef ),
+			mid_header_rewrite     => $self->paramAsScalar( 'ds.mid_header_rewrite', undef ),
+			tr_response_headers    => $self->paramAsScalar( 'ds.tr_response_headers', undef ),
+			tr_request_headers     => $self->paramAsScalar( 'ds.tr_request_headers', undef ),
+			regex_remap        => $self->paramAsScalar( 'ds.regex_remap',        undef ),
+			origin_shield      => $self->paramAsScalar( 'ds.origin_shield',      undef ),
+			cacheurl           => $self->paramAsScalar( 'ds.cacheurl',           undef ),
+			remap_text         => $self->paramAsScalar( 'ds.remap_text',         undef ),
+			initial_dispersion => $self->paramAsScalar( 'ds.initial_dispersion', 1 ),
 		);
 
 		if ( $self->paramAsScalar('ds.type') == &type_id( $self, "DNS" ) ) {
-			$hash{dns_bypass_ip}  = $self->paramAsScalar('ds.dns_bypass_ip');
-			$hash{dns_bypass_ip6} = $self->paramAsScalar('ds.dns_bypass_ip6');
-			$hash{dns_bypass_cname}
-				= $self->paramAsScalar('ds.dns_bypass_cname');
-			$hash{max_dns_answers}
-				= $self->paramAsScalar('ds.max_dns_answers');
-			$hash{dns_bypass_ttl}
-				= $self->paramAsScalar('ds.dns_bypass_ttl') eq ""
+			$hash{dns_bypass_ip}    = $self->paramAsScalar('ds.dns_bypass_ip');
+			$hash{dns_bypass_ip6}   = $self->paramAsScalar('ds.dns_bypass_ip6');
+			$hash{dns_bypass_cname} = $self->paramAsScalar('ds.dns_bypass_cname');
+			$hash{max_dns_answers}  = $self->paramAsScalar('ds.max_dns_answers');
+			$hash{dns_bypass_ttl} =
+				$self->paramAsScalar('ds.dns_bypass_ttl') eq ""
 				? undef
 				: $self->paramAsScalar('ds.dns_bypass_ttl');
 		}
@@ -923,21 +746,15 @@ sub update {
 		}
 
 		# print Dumper( \%hash );
-		my $update
-			= $self->db->resultset('Deliveryservice')->find( { id => $id } );
+		my $update = $self->db->resultset('Deliveryservice')->find( { id => $id } );
 		$update->update( \%hash );
 		$update->update();
-		&log(
-			$self,
-			"Update deliveryservice with xml_id:" . $self->param('ds.xml_id'),
-			"UICHANGE"
-		);
+		&log( $self, "Update deliveryservice with xml_id:" . $self->param('ds.xml_id'), "UICHANGE" );
 
 		# get the existing regexp set in a hash
 		my $regexp_set;
-		my $i  = 0;
-		my $rs = $self->db->resultset('RegexesForDeliveryService')
-			->search( {}, { bind => [$id] } );
+		my $i = 0;
+		my $rs = $self->db->resultset('RegexesForDeliveryService')->search( {}, { bind => [$id] } );
 		while ( my $row = $rs->next ) {
 			$regexp_set->{ $row->id }->{pattern}    = $row->pattern;
 			$regexp_set->{ $row->id }->{type}       = $row->type;
@@ -954,18 +771,15 @@ sub update {
 				if ( defined( $regexp_set->{$re_id} ) ) {
 					my $regexp = $self->param($re_str);
 
-					my $update = $self->db->resultset('Regex')
-						->find( { id => $re_id } );
+					my $update = $self->db->resultset('Regex')->find( { id => $re_id } );
 					$update->update(
-						{   pattern => $regexp,
-							type =>
-								&type_id( $self, $self->param($type_str) ),
+						{
+							pattern => $regexp,
+							type    => &type_id( $self, $self->param($type_str) ),
 						}
 					);
-					$update = $self->db->resultset('DeliveryserviceRegex')
-						->find( { deliveryservice => $id, regex => $re_id } );
-					$update->update(
-						{ set_number => $self->param($set_number_str) } );
+					$update = $self->db->resultset('DeliveryserviceRegex')->find( { deliveryservice => $id, regex => $re_id } );
+					$update->update( { set_number => $self->param($set_number_str) } );
 					$regexp_set->{$re_id}->{updated} = 1;
 				}
 			}
@@ -975,27 +789,25 @@ sub update {
 				my $type_str       = 're_type_new_' . $1;
 				my $re_str         = 're_re_new_' . $1;
 				my $set_number_str = 're_order_new_' . $1;
-				my $type
-					= $self->db->resultset('Type')
-					->search( { name => $self->param($type_str) } )
-					->get_column('id')->single();
-				my $regexp = $self->param($re_str);
+				my $type           = $self->db->resultset('Type')->search( { name => $self->param($type_str) } )->get_column('id')->single();
+				my $regexp         = $self->param($re_str);
 
 				my $insert = $self->db->resultset('Regex')->create(
-					{   pattern => $regexp,
+					{
+						pattern => $regexp,
 						type    => &type_id( $self, $self->param($type_str) ),
 					}
 				);
 				$insert->insert();
 				my $new_re_id = $insert->id;
 
-				my $de_re_insert
-					= $self->db->resultset('DeliveryserviceRegex')->create(
-					{   regex           => $new_re_id,
+				my $de_re_insert = $self->db->resultset('DeliveryserviceRegex')->create(
+					{
+						regex           => $new_re_id,
 						deliveryservice => $id,
 						set_number      => $self->param($set_number_str),
 					}
-					);
+				);
 				$de_re_insert->insert();
 			}
 		}
@@ -1003,52 +815,28 @@ sub update {
 		foreach my $re_id ( keys %{$regexp_set} ) {
 			if ( !defined( $regexp_set->{$re_id}->{updated} ) ) {
 
-				my $delete_re = $self->db->resultset('Regex')
-					->search( { id => $re_id } );
+				my $delete_re = $self->db->resultset('Regex')->search( { id => $re_id } );
 				$delete_re->delete();
 			}
 		}
 
-		$self->header_rewrite(
-			$self->param('id'),
-			$self->param('ds.profile'),
-			$self->param('ds.xml_id'),
-			$self->param('ds.edge_header_rewrite'), "edge"
-		);
-		$self->header_rewrite(
-			$self->param('id'),
-			$self->param('ds.profile'),
-			$self->param('ds.xml_id'),
-			$self->param('ds.mid_header_rewrite'), "mid"
-		);
-		$self->regex_remap(
-			$self->param('id'),        $self->param('ds.profile'),
-			$self->param('ds.xml_id'), $self->param('ds.regex_remap')
-		);
-		$self->cacheurl(
-			$self->param('id'),        $self->param('ds.profile'),
-			$self->param('ds.xml_id'), $self->param('ds.cacheurl')
-		);
+		$self->header_rewrite( $self->param('id'), $self->param('ds.profile'), $self->param('ds.xml_id'), $self->param('ds.edge_header_rewrite'), "edge" );
+		$self->header_rewrite( $self->param('id'), $self->param('ds.profile'), $self->param('ds.xml_id'), $self->param('ds.mid_header_rewrite'),  "mid" );
+		$self->regex_remap( $self->param('id'), $self->param('ds.profile'), $self->param('ds.xml_id'), $self->param('ds.regex_remap') );
+		$self->cacheurl( $self->param('id'), $self->param('ds.profile'), $self->param('ds.xml_id'), $self->param('ds.cacheurl') );
 
 		$self->flash( message => "Delivery service updated!" );
 		return $self->redirect_to( '/ds/' . $id );
 	}
 	else {
 		&stash_role($self);
-		my $rs_ds
-			= $self->db->resultset('Deliveryservice')
-			->search( { 'me.id' => $id },
-			{ prefetch => [ { 'type' => undef }, { 'profile' => undef } ] } );
-		my $data         = $rs_ds->single;
+		my $rs_ds = $self->db->resultset('Deliveryservice')->search( { 'me.id' => $id }, { prefetch => [ { 'type' => undef }, { 'profile' => undef } ] } );
+		my $data = $rs_ds->single;
 		my $cdn_domain   = &get_cdn_domain( $self, $id );
-		my $server_count = $self->db->resultset('DeliveryserviceServer')
-			->search( { deliveryservice => $id } )->count();
-		my $static_count = $self->db->resultset('Staticdnsentry')
-			->search( { deliveryservice => $id } )->count();
-		my $regexp_set = &get_regexp_set( $self, $id );
-		my @example_urls
-			= &get_example_urls( $self, $id, $regexp_set, $data, $cdn_domain,
-			$data->protocol );
+		my $server_count = $self->db->resultset('DeliveryserviceServer')->search( { deliveryservice => $id } )->count();
+		my $static_count = $self->db->resultset('Staticdnsentry')->search( { deliveryservice => $id } )->count();
+		my $regexp_set   = &get_regexp_set( $self, $id );
+		my @example_urls = &get_example_urls( $self, $id, $regexp_set, $data, $cdn_domain, $data->protocol );
 		my $action;
 
 		$self->stash(
@@ -1071,13 +859,11 @@ sub check_regexp {
 	my $sep = "__NEWLINE__";
 	my $err = "";
 	if ( $re =~ /^\s/ || $re =~ /\s$/ ) {
-		$err
-			.= "Regular expression can not start or end with whitespace... (did you cut-and-paste?)";
+		$err .= "Regular expression can not start or end with whitespace... (did you cut-and-paste?)";
 	}
 
 	if ( $re eq ".*" ) {
-		$err
-			= ".* is not a valid regexp, at least not for any of the CDN entries.";
+		$err = ".* is not a valid regexp, at least not for any of the CDN entries.";
 	}
 
 	my $regexp_err = &is_regexp($re);
@@ -1105,71 +891,53 @@ sub create {
 
 	if ( $self->check_deliveryservice_input() ) {
 		my $insert = $self->db->resultset('Deliveryservice')->create(
-			{   xml_id         => $self->paramAsScalar('ds.xml_id'),
-				display_name   => $self->paramAsScalar('ds.display_name'),
-				dscp           => $self->paramAsScalar( 'ds.dscp', 0 ),
-				signed         => $self->paramAsScalar('ds.signed'),
-				qstring_ignore => $self->paramAsScalar('ds.qstring_ignore'),
-				geo_limit      => $self->paramAsScalar('ds.geo_limit'),
-				http_bypass_fqdn =>
-					$self->paramAsScalar('ds.http_bypass_fqdn'),
-				dns_bypass_ip  => $self->paramAsScalar('ds.dns_bypass_ip'),
-				dns_bypass_ip6 => $self->paramAsScalar('ds.dns_bypass_ip6'),
-				dns_bypass_cname =>
-					$self->paramAsScalar('ds.dns_bypass_cname'),
-				dns_bypass_ttl  => $self->paramAsScalar('ds.dns_bypass_ttl'),
-				org_server_fqdn => $self->paramAsScalar('ds.org_server_fqdn'),
-				multi_site_origin =>
-					$self->paramAsScalar('ds.multi_site_origin'),
-				ccr_dns_ttl     => $self->paramAsScalar('ds.ccr_dns_ttl'),
-				type            => $self->paramAsScalar('ds.type'),
-				cdn_id          => $cdn_id,
-				profile         => $self->paramAsScalar('ds.profile'),
-				global_max_mbps => $self->hr_string_to_mbps(
-					$self->paramAsScalar( 'ds.global_max_mbps', 0 )
-				),
-				global_max_tps =>
-					$self->paramAsScalar( 'ds.global_max_tps', 0 ),
-				miss_lat    => $self->paramAsScalar('ds.miss_lat'),
-				miss_long   => $self->paramAsScalar('ds.miss_long'),
-				long_desc   => $self->paramAsScalar('ds.long_desc'),
-				long_desc_1 => $self->paramAsScalar('ds.long_desc_1'),
-				long_desc_2 => $self->paramAsScalar('ds.long_desc_2'),
-				max_dns_answers =>
-					$self->paramAsScalar( 'ds.max_dns_answers', 0 ),
-				info_url   => $self->paramAsScalar('ds.info_url'),
-				check_path => $self->paramAsScalar('ds.check_path'),
+			{
+				xml_id                 => $self->paramAsScalar('ds.xml_id'),
+				display_name           => $self->paramAsScalar('ds.display_name'),
+				dscp                   => $self->paramAsScalar( 'ds.dscp', 0 ),
+				signed                 => $self->paramAsScalar('ds.signed'),
+				qstring_ignore         => $self->paramAsScalar('ds.qstring_ignore'),
+				geo_limit              => $self->paramAsScalar('ds.geo_limit'),
+				http_bypass_fqdn       => $self->paramAsScalar('ds.http_bypass_fqdn'),
+				dns_bypass_ip          => $self->paramAsScalar('ds.dns_bypass_ip'),
+				dns_bypass_ip6         => $self->paramAsScalar('ds.dns_bypass_ip6'),
+				dns_bypass_cname       => $self->paramAsScalar('ds.dns_bypass_cname'),
+				dns_bypass_ttl         => $self->paramAsScalar('ds.dns_bypass_ttl'),
+				org_server_fqdn        => $self->paramAsScalar('ds.org_server_fqdn'),
+				multi_site_origin      => $self->paramAsScalar('ds.multi_site_origin'),
+				ccr_dns_ttl            => $self->paramAsScalar('ds.ccr_dns_ttl'),
+				type                   => $self->paramAsScalar('ds.type'),
+				cdn_id                 => $cdn_id,
+				profile                => $self->paramAsScalar('ds.profile'),
+				global_max_mbps        => $self->hr_string_to_mbps( $self->paramAsScalar( 'ds.global_max_mbps', 0 ) ),
+				global_max_tps         => $self->paramAsScalar( 'ds.global_max_tps', 0 ),
+				miss_lat               => $self->paramAsScalar('ds.miss_lat'),
+				miss_long              => $self->paramAsScalar('ds.miss_long'),
+				long_desc              => $self->paramAsScalar('ds.long_desc'),
+				long_desc_1            => $self->paramAsScalar('ds.long_desc_1'),
+				long_desc_2            => $self->paramAsScalar('ds.long_desc_2'),
+				max_dns_answers        => $self->paramAsScalar( 'ds.max_dns_answers', 0 ),
+				info_url               => $self->paramAsScalar('ds.info_url'),
+				check_path             => $self->paramAsScalar('ds.check_path'),
 				regional_geo_blocking  => $self->paramAsScalar('ds.regional_geo_blocking'),
-				active     => $self->paramAsScalar('ds.active'),
-				protocol   => $self->paramAsScalar('ds.protocol'),
-				ipv6_routing_enabled =>
-					$self->paramAsScalar('ds.ipv6_routing_enabled'),
-				range_request_handling =>
-					$self->paramAsScalar('ds.range_request_handling'),
-				edge_header_rewrite =>
-					$self->paramAsScalar('ds.edge_header_rewrite'),
-				mid_header_rewrite =>
-					$self->paramAsScalar( 'ds.mid_header_rewrite', undef ),
-				regex_remap =>
-					$self->paramAsScalar( 'ds.regex_remap', undef ),
-				origin_shield =>
-					$self->paramAsScalar( 'ds.origin_shield', undef ),
-				cacheurl   => $self->paramAsScalar( 'ds.cacheurl',   undef ),
-				remap_text => $self->paramAsScalar( 'ds.remap_text', undef ),
-				initial_dispersion =>
-					$self->paramAsScalar( 'ds.initial_dispersion', 1 ),
+				active                 => $self->paramAsScalar('ds.active'),
+				protocol               => $self->paramAsScalar('ds.protocol'),
+				ipv6_routing_enabled   => $self->paramAsScalar('ds.ipv6_routing_enabled'),
+				range_request_handling => $self->paramAsScalar('ds.range_request_handling'),
+				edge_header_rewrite    => $self->paramAsScalar('ds.edge_header_rewrite'),
+				mid_header_rewrite     => $self->paramAsScalar( 'ds.mid_header_rewrite', undef ),
+				regex_remap        => $self->paramAsScalar( 'ds.regex_remap',        undef ),
+				origin_shield      => $self->paramAsScalar( 'ds.origin_shield',      undef ),
+				cacheurl           => $self->paramAsScalar( 'ds.cacheurl',           undef ),
+				remap_text         => $self->paramAsScalar( 'ds.remap_text',         undef ),
+				initial_dispersion => $self->paramAsScalar( 'ds.initial_dispersion', 1 ),
 			}
 		);
 		$insert->insert();
 		$new_id = $insert->id;
-		&log(
-			$self,
-			"Create deliveryservice with xml_id:" . $self->param('ds.xml_id'),
-			"UICHANGE"
-		);
+		&log( $self, "Create deliveryservice with xml_id:" . $self->param('ds.xml_id'), "UICHANGE" );
 
-		if ( $new_id == -1 )
-		{    # there was an error the flash will already be set,
+		if ( $new_id == -1 ) {    # there was an error the flash will already be set,
 			my $referer = $self->req->headers->header('referer');
 			my $qstring = "?";
 			my @params  = $self->param;
@@ -1208,63 +976,40 @@ sub create {
 			if ( !defined( $re->{order} ) ) {
 				next;
 			}    # 0 gets iterated over if the form sends just a _1
-			my $type
-				= $self->db->resultset('Type')
-				->search( { name => $re->{type} } )->get_column('id')
-				->single();
+			my $type = $self->db->resultset('Type')->search( { name => $re->{type} } )->get_column('id')->single();
 			my $regexp = $re->{re};
 
 			my $insert = $self->db->resultset('Regex')->create(
-				{   pattern => $regexp,
+				{
+					pattern => $regexp,
 					type    => $type,
 				}
 			);
 			$insert->insert();
 			my $new_re_id = $insert->id;
 
-			my $de_re_insert
-				= $self->db->resultset('DeliveryserviceRegex')->create(
-				{   regex           => $new_re_id,
+			my $de_re_insert = $self->db->resultset('DeliveryserviceRegex')->create(
+				{
+					regex           => $new_re_id,
 					deliveryservice => $new_id,
 					set_number      => $re->{order},
 				}
-				);
+			);
 			$de_re_insert->insert();
 		}
 
-		$self->header_rewrite(
-			$new_id,
-			$self->param('ds.profile'),
-			$self->param('ds.xml_id'),
-			$self->param('ds.edge_header_rewrite'), "edge"
-		);
-		$self->header_rewrite(
-			$new_id,
-			$self->param('ds.profile'),
-			$self->param('ds.xml_id'),
-			$self->param('ds.mid_header_rewrite'), "mid"
-		);
-		$self->regex_remap(
-			$self->param('ds.profile'),
-			$self->param('ds.xml_id'),
-			$self->param('ds.regex_remap')
-		);
-		$self->cacheurl(
-			$new_id,
-			$self->param('ds.profile'),
-			$self->param('ds.xml_id'),
-			$self->param('ds.cacheurl')
-		);
+		$self->header_rewrite( $new_id, $self->param('ds.profile'), $self->param('ds.xml_id'), $self->param('ds.edge_header_rewrite'), "edge" );
+		$self->header_rewrite( $new_id, $self->param('ds.profile'), $self->param('ds.xml_id'), $self->param('ds.mid_header_rewrite'),  "mid" );
+		$self->regex_remap( $self->param('ds.profile'), $self->param('ds.xml_id'), $self->param('ds.regex_remap') );
+		$self->cacheurl( $new_id, $self->param('ds.profile'), $self->param('ds.xml_id'), $self->param('ds.cacheurl') );
 
 		##create dnssec keys for the new DS if DNSSEC is enabled for the CDN
-		my $cdn_rs = $self->db->resultset('Cdn')->search( { id => $cdn_id } )
-			->single();
+		my $cdn_rs = $self->db->resultset('Cdn')->search( { id => $cdn_id } )->single();
 		my $dnssec_enabled = $cdn_rs->dnssec_enabled;
 
 		if ( $dnssec_enabled == 1 ) {
 			$self->app->log->debug("dnssec is enabled, creating dnssec keys");
-			$self->create_dnssec_keys( $cdn_rs->name,
-				$self->param('ds.xml_id'), $new_id );
+			$self->create_dnssec_keys( $cdn_rs->name, $self->param('ds.xml_id'), $new_id );
 		}
 		$self->flash( message => "Success!" );
 		return $self->redirect_to( '/ds/' . $new_id );
@@ -1298,7 +1043,7 @@ sub create_dnssec_keys {
 	my $get_keys = $response_container->{'response'};
 	$keys = decode_json( $get_keys->content );
 
-#get default expiration days and ttl for DSs from CDN record to use when generating new keys
+	#get default expiration days and ttl for DSs from CDN record to use when generating new keys
 	my $cdn_ksk = $keys->{$cdn_name}->{ksk};
 	my $k_exp_days = $self->get_key_expiration_days( $cdn_ksk, "365" );
 
@@ -1310,15 +1055,12 @@ sub create_dnssec_keys {
 	#create the ds domain name for dnssec keys
 	my $domain_name             = $self->get_cdn_domain($ds_id);
 	my $deliveryservice_regexes = $self->get_regexp_set($ds_id);
-	my $rs_ds
-		= $self->db->resultset('Deliveryservice')
-		->search( { 'me.xml_id' => $xml_id },
-		{ prefetch => [ { 'type' => undef }, { 'profile' => undef } ] } );
-	my $data         = $rs_ds->single;
-	my @example_urls = UI::DeliveryService::get_example_urls( $self, $ds_id,
-		$deliveryservice_regexes, $data, $domain_name, $data->protocol );
+	my $rs_ds =
+		$self->db->resultset('Deliveryservice')->search( { 'me.xml_id' => $xml_id }, { prefetch => [ { 'type' => undef }, { 'profile' => undef } ] } );
+	my $data = $rs_ds->single;
+	my @example_urls = UI::DeliveryService::get_example_urls( $self, $ds_id, $deliveryservice_regexes, $data, $domain_name, $data->protocol );
 
-#first one is the one we want.  period at end for dnssec, substring off stuff we dont want
+	#first one is the one we want.  period at end for dnssec, substring off stuff we dont want
 	my $ds_name = $example_urls[0] . ".";
 	my $length = length($ds_name) - CORE::index( $ds_name, "." );
 	$ds_name = substr( $ds_name, CORE::index( $ds_name, "." ) + 1, $length );
@@ -1327,12 +1069,8 @@ sub create_dnssec_keys {
 	my $z_expiration = $inception + ( 86400 * $z_exp_days );
 	my $k_expiration = $inception + ( 86400 * $k_exp_days );
 
-	my $zsk
-		= $self->get_dnssec_keys( "zsk", $ds_name, $dnskey_ttl, $inception,
-		$z_expiration, "new", $inception );
-	my $ksk
-		= $self->get_dnssec_keys( "ksk", $ds_name, $dnskey_ttl, $inception,
-		$k_expiration, "new", $inception );
+	my $zsk = $self->get_dnssec_keys( "zsk", $ds_name, $dnskey_ttl, $inception, $z_expiration, "new", $inception );
+	my $ksk = $self->get_dnssec_keys( "ksk", $ds_name, $dnskey_ttl, $inception, $k_expiration, "new", $inception );
 
 	#add to keys hash
 	$keys->{$xml_id} = {
@@ -1351,7 +1089,7 @@ sub get_key_expiration_days {
 	my $default_exp = shift;
 	foreach my $key (@$keys) {
 		my $status = $key->{status};
-		if ( $status eq 'new' ) { #ignore anything other than the 'new' record
+		if ( $status eq 'new' ) {    #ignore anything other than the 'new' record
 			my $exp   = $key->{expirationDate};
 			my $incep = $key->{inceptionDate};
 			return ( $exp - $incep ) / 86400;
@@ -1367,7 +1105,7 @@ sub get_key_ttl {
 
 	foreach my $key (@$keys) {
 		my $status = $key->{status};
-		if ( $status eq 'new' ) { #ignore anything other than the 'new' record
+		if ( $status eq 'new' ) {    #ignore anything other than the 'new' record
 			return $key->{ttl};
 		}
 	}

--- a/traffic_ops/app/lib/UI/DeliveryService.pm
+++ b/traffic_ops/app/lib/UI/DeliveryService.pm
@@ -298,7 +298,7 @@ sub check_deliveryservice_input {
 
 	# TODO:  what restrictions on display_name?
 
-	my $typename = $self->db->resultset('Type')->search( { id => $self->param('ds.type') } )->get_column('name')->single();
+	my $typename = $self->param('ds.type.name');
 	if ( $typename eq 'ANY_MAP' ) {
 		return $self->valid;    # Anything goes for the ANY_MAP, but ds.type is only set on create
 	}
@@ -682,7 +682,6 @@ sub delete_cfg_file {
 sub update {
 	my $self = shift;
 	my $id   = $self->param('id');
-
 	if ( !&is_oper($self) ) {
 		my $err = "You do not have enough privileges to modify this.";
 		$self->flash( message => $err );
@@ -703,7 +702,7 @@ sub update {
 			org_server_fqdn        => $self->paramAsScalar('ds.org_server_fqdn'),
 			multi_site_origin      => $self->paramAsScalar('ds.multi_site_origin'),
 			ccr_dns_ttl            => $self->paramAsScalar('ds.ccr_dns_ttl'),
-			type                   => $self->paramAsScalar('ds.type'),
+			type                   => $self->paramAsScalar('ds.type.id'),
 			cdn_id                 => $self->paramAsScalar('ds.cdn_id'),
 			profile                => $self->paramAsScalar('ds.profile'),
 			global_max_mbps        => $self->hr_string_to_mbps( $self->paramAsScalar( 'ds.global_max_mbps', 0 ) ),
@@ -731,7 +730,7 @@ sub update {
 			initial_dispersion => $self->paramAsScalar( 'ds.initial_dispersion', 1 ),
 		);
 
-		if ( $self->paramAsScalar('ds.type') == &type_id( $self, "DNS" ) ) {
+		if ( $self->paramAsScalar('ds.type.name') eq "DNS" ) {
 			$hash{dns_bypass_ip}    = $self->paramAsScalar('ds.dns_bypass_ip');
 			$hash{dns_bypass_ip6}   = $self->paramAsScalar('ds.dns_bypass_ip6');
 			$hash{dns_bypass_cname} = $self->paramAsScalar('ds.dns_bypass_cname');

--- a/traffic_ops/app/lib/UI/DeliveryService.pm
+++ b/traffic_ops/app/lib/UI/DeliveryService.pm
@@ -289,6 +289,18 @@ sub delete {
 	return $self->redirect_to('/close_fancybox.html');
 }
 
+sub typeid {
+	my $self = shift;
+	print STDOUT "typeid: " . Dumper( { map { $_ => $self->param($_) } grep /ds.type/, $self->param() } ) . "\n";
+	return $self->param('ds.type.id') // $self->param('ds.type');
+}
+
+sub typename {
+	my $self = shift;
+	print STDOUT "typename: " . Dumper( { map { $_ => $self->param($_) } grep /ds.type/, $self->param() } ) . "\n";
+	return $self->param('type.name') // $self->db->resultset('Type')->search( { id => $self->typeid() } )->get_column('name')->single();
+}
+
 sub check_deliveryservice_input {
 	my $self = shift;
 
@@ -298,7 +310,7 @@ sub check_deliveryservice_input {
 
 	# TODO:  what restrictions on display_name?
 
-	my $typename = $self->param('ds.type.name');
+	my $typename = $self->typename();
 	if ( $typename eq 'ANY_MAP' ) {
 		return $self->valid;    # Anything goes for the ANY_MAP, but ds.type is only set on create
 	}
@@ -730,7 +742,8 @@ sub update {
 			initial_dispersion => $self->paramAsScalar( 'ds.initial_dispersion', 1 ),
 		);
 
-		if ( $self->paramAsScalar('ds.type.name') eq "DNS" ) {
+		my $typename = $self->typename();
+		if ( $typename eq "DNS" ) {
 			$hash{dns_bypass_ip}    = $self->paramAsScalar('ds.dns_bypass_ip');
 			$hash{dns_bypass_ip6}   = $self->paramAsScalar('ds.dns_bypass_ip6');
 			$hash{dns_bypass_cname} = $self->paramAsScalar('ds.dns_bypass_cname');

--- a/traffic_ops/app/lib/UI/DeliveryService.pm
+++ b/traffic_ops/app/lib/UI/DeliveryService.pm
@@ -291,13 +291,11 @@ sub delete {
 
 sub typeid {
 	my $self = shift;
-	print STDOUT "typeid: " . Dumper( { map { $_ => $self->param($_) } grep /ds.type/, $self->param() } ) . "\n";
 	return $self->param('ds.type.id') // $self->param('ds.type');
 }
 
 sub typename {
 	my $self = shift;
-	print STDOUT "typename: " . Dumper( { map { $_ => $self->param($_) } grep /ds.type/, $self->param() } ) . "\n";
 	return $self->param('type.name') // $self->db->resultset('Type')->search( { id => $self->typeid() } )->get_column('name')->single();
 }
 
@@ -714,7 +712,7 @@ sub update {
 			org_server_fqdn        => $self->paramAsScalar('ds.org_server_fqdn'),
 			multi_site_origin      => $self->paramAsScalar('ds.multi_site_origin'),
 			ccr_dns_ttl            => $self->paramAsScalar('ds.ccr_dns_ttl'),
-			type                   => $self->paramAsScalar('ds.type.id'),
+			type                   => $self->typeid(),
 			cdn_id                 => $self->paramAsScalar('ds.cdn_id'),
 			profile                => $self->paramAsScalar('ds.profile'),
 			global_max_mbps        => $self->hr_string_to_mbps( $self->paramAsScalar( 'ds.global_max_mbps', 0 ) ),

